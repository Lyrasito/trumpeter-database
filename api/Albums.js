const express = require("express");
const albumsRouter = express.Router({ mergeParams: true });
const sqlite3 = require("sqlite3");
const db = new sqlite3.Database("./database.sqlite");

//Get all albums by one Player
albumsRouter.get("/", (req, res, next) => {
  db.all(
    `SELECT * FROM 'Album' WHERE Album.player_id = ${req.player.id}`,
    (err, albums) => {
      if (err) {
        next(err);
      } else {
        res.send({ albums: albums });
      }
    }
  );
});
//get all genres from one Player
albumsRouter.get("/genres", (req, res, next) => {
  db.all(
    `SELECT * FROM 'Album' WHERE Album.player_id = ${req.player.id}`,
    (err, albums) => {
      if (err) {
        next(err);
      } else {
        const genres = albums.map((album) => {
          return album.genre;
        });
        uniqueArray = genres.filter(function (item, pos, self) {
          return self.indexOf(item) == pos;
        });
        res.send({ genre: uniqueArray });
      }
    }
  );
});

module.exports = albumsRouter;
