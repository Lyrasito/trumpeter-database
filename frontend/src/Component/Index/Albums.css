.albumFilter {
  display: flex;
  float: right;
  padding: 10px;
}

.albumList {
  display: flex;
  flex-wrap: wrap;
  width: 100%;
}
