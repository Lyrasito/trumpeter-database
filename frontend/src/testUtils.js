const fakePlayer = () => ({
  id: 1,
  name: "Fake Player",
  city: "Fake City",
  startYear: 1234,
  endYear: 5678,
});

const fakePlayers = () => [
  {
    id: 1,
    name: "Fake Player",
    city: "Fake City",
    startYear: 1234,
    endYear: 5678,
  },
  {
    id: 2,
    name: "Fake Player 2",
    city: "Fake City 2",
    startYear: 2345,
    endYear: 6789,
  },
];

const fakeAlbums = () => [
  {
    id: 1,
    title: "Fake Album 1",
    year: 1234,
    genre: "fakeGenre1",
    player_id: 1,
  },
  {
    id: 2,
    title: "Fake Album 2",
    year: 1234,
    genre: "fakeGenre2",
    player_id: 1,
  },
];

const fakeAlbum = () => ({
  id: 1,
  title: "Fake Album 1",
  year: 1234,
  genre: "fakeGenre1",
  player_id: 1,
});

const fakeGenres = () => ["fakeGenre1", "fakeGenre2"];

export { fakeAlbum, fakeAlbums, fakePlayer, fakeGenres, fakePlayers };
