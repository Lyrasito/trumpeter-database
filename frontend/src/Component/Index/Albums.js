//props: album, player

import React from "react";
import "./Albums.css";
import Spotify from "../../Spotify";
import Album from "./Album";

class Albums extends React.Component {
  state = {
    albumLinks: [],
    genre: "",
  };

  componentDidMount() {
    this.getAlbumLinks();
  }

  getSpotify = async (term) => {
    const link = await Spotify.searchAlbum(term);
    return link;
  };

  getAlbumLinks = async () => {
    for (let i = 0; i < this.props.albums.length; i++) {
      let link = await this.getSpotify(
        `${this.props.albums[i].title} ${this.props.player.name}`
      );
      let links = this.state.albumLinks;
      links.push(link);
      this.setState({
        albumLinks: links,
      });
    }
  };

  getGenre = (event) => {
    this.setState({ genre: event.target.value });
  };
  filterByGenre = () => {
    this.props.filterByGenre(this.state.genre);
  };
  render() {
    return (
      <div id="shownAlbums">
        <div className="albumFilter">
          <label>Filter by genre:</label>
          <select className="genreSelect" onChange={this.getGenre}>
            <option value="noFilter">No filter</option>
            {this.props.player.genres.map((genre, index) => {
              return (
                <option value={genre} key={index}>
                  {genre}
                </option>
              );
            })}
          </select>
          <button
            type="submit"
            className="filterButton"
            onClick={this.filterByGenre}
          >
            Filter
          </button>
        </div>
        <div className="albumList">
          {this.props.albums.map((album, index) => (
            <Album
              album={album}
              albumLink={this.state.albumLinks[index]}
              key={album.id}
            />
          ))}
        </div>
      </div>
    );
  }
}

export default Albums;

/*
 */
