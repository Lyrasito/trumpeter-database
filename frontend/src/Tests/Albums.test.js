import Albums from "../Component/Index/Albums";
import React from "react";
import toJSON from "enzyme-to-json";
import { shallow } from "enzyme";
import { fakePlayer, fakeAlbums, fakeGenres } from "../testUtils";

const player = fakePlayer();
const genres = fakeGenres();
const albums = fakeAlbums();

describe("<Albums />", () => {
  it("renders properly", () => {
    const wrapper = shallow(
      <Albums albums={albums} genres={genres} player={player} />
    );
    expect(wrapper.find("Album")).toHaveLength(2);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
  it("filters rendered <Album />s by genre and removes the filter if 'No filter' is selected", async () => {
    const wrapper = shallow(
      <Albums albums={albums} genres={genres} player={player} />
    );
    const filter = wrapper.find(".genreSelect");
    filter.simulate("change", { target: { value: genres[0] } });
    const button = wrapper.find(".filterButton");
    button.simulate("click");
    const album = wrapper.find("Album");
    expect(album).toHaveLength(1);
    expect(album.props().album.genre).toEqual(genres[0]);

    filter.simulate("change", { target: { value: "noFilter" } });
    button.simulate("click");
    const album1 = wrapper.find("Album");
    expect(album1).toHaveLength(2);
  });
});
