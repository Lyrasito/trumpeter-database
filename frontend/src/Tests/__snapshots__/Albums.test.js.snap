// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Albums /> renders properly 1`] = `
<div
  id="shownAlbums"
>
  <div
    className="albumFilter"
  >
    <label>
      Filter by genre:
    </label>
    <select
      className="genreSelect"
      onChange={[Function]}
    >
      <option
        value="noFilter"
      >
        No filter
      </option>
      <option
        key="0"
        value="fakeGenre1"
      >
        fakeGenre1
      </option>
      <option
        key="1"
        value="fakeGenre2"
      >
        fakeGenre2
      </option>
    </select>
    <button
      className="filterButton"
      onClick={[Function]}
      type="submit"
    >
      Filter
    </button>
  </div>
  <div
    className="albumList"
  >
    <Album
      album={
        Object {
          "genre": "fakeGenre1",
          "id": 1,
          "player_id": 1,
          "title": "Fake Album 1",
          "year": 1234,
        }
      }
      key="1"
      player={
        Object {
          "city": "Fake City",
          "endYear": 5678,
          "id": 1,
          "name": "Fake Player",
          "startYear": 1234,
        }
      }
    />
    <Album
      album={
        Object {
          "genre": "fakeGenre2",
          "id": 2,
          "player_id": 1,
          "title": "Fake Album 2",
          "year": 1234,
        }
      }
      key="2"
      player={
        Object {
          "city": "Fake City",
          "endYear": 5678,
          "id": 1,
          "name": "Fake Player",
          "startYear": 1234,
        }
      }
    />
  </div>
</div>
`;
